James 1874 Various optimisation for JMAP

chibenwa opened the pull request:
This PR aims at minimising the number of calls to an external DB. Thus : 
 - We can retrieve all mailboxCounters once as they are collocated on the same row
 - We can avoid a read per mailboxes for GetMailboxes as we can retrieve parentId from context

Then (not related to @ddolcimascolo complains bu still cool IMO ) : 
 - Avoid a read on an index by making MailboxManager::save return the MailboxId. This cleans the code of MailboxFactory btw. I provide more tests to be confident with the code of MailboxPath.

chibenwa commented:
```
[Building Java 8 james-1874-optim-91a72bd5-f72c-4ad2-b8ae-b3de5eb3d8db] Tests in error: 
[Building Java 8 james-1874-optim-91a72bd5-f72c-4ad2-b8ae-b3de5eb3d8db]   DefaultMailboxesProvisioningFilterThreadTest.testConcurrentAccessToFilterShouldNotThrow:87 » Runtime
[Building Java 8 james-1874-optim-91a72bd5-f72c-4ad2-b8ae-b3de5eb3d8db] 
[Building Java 8 james-1874-optim-91a72bd5-f72c-4ad2-b8ae-b3de5eb3d8db] Tests run: 637, Failures: 0, Errors: 1, Skipped: 2
```

Will also test on my local...

chibenwa commented:
test this please

mbaechler commented on mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxCounters.java:
```diff
@@ -0,0 +1,56 @@
+    private final long count;
+    private final long unseen;
+
+    public MailboxCounters(long count, long unseen) {
```
could you create a builder to avoid such ambiguous constructor ?

mbaechler commented on mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java:
```diff
@@ -65,6 +66,7 @@
 
 public class CassandraMessageMapper implements MessageMapper {
     private static final Logger LOGGER = LoggerFactory.getLogger(CassandraMessageMapper.class);
+    public static final MailboxCounters EMPTY_COUNTERS = new MailboxCounters(0L, 0L);
```
s/empty/initial/ ?

mbaechler commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java:
```diff
@@ -102,17 +101,19 @@ private GetMailboxesResponse getMailboxesResponse(GetMailboxesRequest mailboxesR
     }
 
-    private Stream<MailboxId> retrieveMailboxIds(Optional<ImmutableList<MailboxId>> mailboxIds, MailboxSession mailboxSession) throws MailboxException {
+    private Stream<Optional<Mailbox>> retrieveMailboxes(Optional<ImmutableList<MailboxId>> mailboxIds, MailboxSession mailboxSession) throws MailboxException {
```
why do you want to change the return type ? Do we care about retrieving absent mailboxes ?

chibenwa replied:
> why do you want to change the return type ?

Because I needed to go one Stream operation further

> Do we care about retrieving absent mailboxes ?

No that's why the caller filter it.

Actually you might know that optinal unboxing is a pain with java8, it's not scala where you can simply do it with flatMap. Here, I let the caller (which is private) do the unboxing for my branching.

mbaechler replied:
private method is not an excuse to be lazy (:
if you don't care about empty element, you have to return a Stream<Mailbox>, doing otherwise is just promoting easy writing over easy reading

chibenwa replied:
I added tooling in java-8-utils OptionalConverter a static method (and related tests) for Optional convertion to stream.

This makes Optional unboxing easy to read, and I applied it on each branch of the if.

mbaechler commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java:
```diff
@@ -102,17 +101,19 @@ private GetMailboxesResponse getMailboxesResponse(GetMailboxesRequest mailboxesR
                 .stream()
-                .map(MailboxMetaData::getId);
+                .map(MailboxMetaData::getId)
+                .map(mailboxId -> mailboxFactory.fromMailboxId(mailboxId, userMailboxes, mailboxSession));
```
it looks like you reached the point where the factory is doing to many things (because method names are hard to find to express that optional second argument).

chibenwa replied:
That's why last commit make the mailboxFactory take only MailboxId.

By the way, which pattern would you propose for this ?

mbaechler replied:
the better solution I see is doing a builder (sorry that you don't have named arguments, yet)

mbaechler commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MailboxFactory.java:
```diff
@@ -93,13 +103,22 @@ public MailboxFactory(MailboxManager mailboxManager) {
         }
         MailboxPath parent = levels.get(levels.size() - 2);
-        return Optional.of(getMailboxId(parent, mailboxSession));
+        if (userMailboxesMetadata.isPresent()) {
```
if on optional is an anti-pattern.

chibenwa replied:
I will see what I can do.

mbaechler commented on mailbox/api/src/test/java/org/apache/james/mailbox/model/MailboxPathTest.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @Test
+    public void getHierarchyLevelsShouldWorkWhenOneLevel() {
```
shouldWork doesn't make sense, you have to describe what is expected

mbaechler commented on mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java:
```diff
@@ -153,7 +155,7 @@
      *             when creation fails
      */
-    void createMailbox(MailboxPath mailboxPath, MailboxSession mailboxSession) throws MailboxException;
+    Optional<MailboxId> createMailbox(MailboxPath mailboxPath, MailboxSession mailboxSession) throws MailboxException;
```
documentation doesn't explain why optional ?
And what about "intermediary mailboxes created" ?

chibenwa replied:
Documentation modified.

mbaechler commented on mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/MaildirMailboxMapper.java:
```diff
@@ -229,7 +229,8 @@ public void save(Mailbox mailbox) throws MailboxException {
             folder.setACL(session, mailbox.getACL());
         }
-        
+        cacheMailbox(mailbox);
```
why ?

chibenwa replied:
cacheMailbox is the operation that add the mailboxId.

It was called on the first message operation on the mailbox.

mbaechler commented on mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java:
```diff
@@ -538,7 +540,12 @@ public void runVoid() throws MailboxException {
 
+            int mailboxIdCount = mailboxIds.size();
+            if (mailboxIdCount > 0) {
+                return Optional.of(mailboxIds.get(mailboxIdCount - 1));
```
do you mean "last" ? There's at least Iterables that provides this

chibenwa replied:
See @mbaechler comment : It is the case, and I updated code to make it blattent.

mbaechler commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java:
```diff
@@ -104,10 +105,12 @@ private void createMailbox(MailboxCreationId mailboxCreationId, MailboxCreateReq
-            subscriptionManager.subscribe(mailboxSession, mailboxPath.getName());
-            Optional<Mailbox> mailbox = mailboxFactory.fromMailboxPath(mailboxPath, mailboxSession);
+            Optional<MailboxId> mailboxId = OptionalConverter.fromGuava(mailboxManager.createMailbox(mailboxPath, mailboxSession));
+            Optional<Mailbox> mailbox;
```
why it's splitted on two lines ?

chibenwa replied:
Because I wanted to extract and forgot to finish. Sorry.

mbaechler commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java:
```diff
@@ -104,10 +105,12 @@ private void createMailbox(MailboxCreationId mailboxCreationId, MailboxCreateReq
+            Optional<MailboxId> mailboxId = OptionalConverter.fromGuava(mailboxManager.createMailbox(mailboxPath, mailboxSession));
+            Optional<Mailbox> mailbox;
+            mailbox = mailboxId.map(id -> mailboxFactory.fromMailboxId(id, mailboxSession))
+                .orElseGet(Optional::empty);
```
you don't need this line

chibenwa replied:
Actually suggested by my IDE. It's the equivalent of a flatMap, absent from optional API.

mbaechler replied:
maybe I don't get it, but http://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#flatMap-java.util.function.Function- says it exists

chibenwa replied:
Yep

chibenwa commented:
@mbaechler Concerning the MailboxFactory stuff, you have two public methods now : 

```
public Optional<Mailbox> fromMailboxId(MailboxId mailboxId, MailboxSession mailboxSession)

public Optional<Mailbox> fromMailboxId(MailboxId mailboxId, List<MailboxMetaData> userMailboxesMetadata, MailboxSession mailboxSession)
```

No conflicting types, only one optional argument, I have troubble to see the benefits of a builder. If any other team member support this clame, I will do it, but I do not see the benefits...

Seems OK to me, no? No need to dig into some builder crazy stuff IMO...

mbaechler commented:
Ok, let me explain the problem : 
`public Optional<Mailbox> fromMailboxId(MailboxId mailboxId, List<MailboxMetaData> userMailboxesMetadata, MailboxSession mailboxSession)` says : I will maybe build a Mailbox based on the MailboxId.
It happens to need a session, like most method, so ok for this one, it's very usual.
But as a caller, I don't understand why I would need to pass a List of MailboxMetaData. The truth is that it replaces the store call, but the design is very unusual and it breaks the "least surprise" principle.
We have to search for naming that describe this behavior, builder or not.
Optimisation should not compromise discoverability of the API IMO.

strokyl commented on mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxCounterDAOTest.java:
```diff
@@ -79,6 +87,34 @@ public void incrementUnseenShouldAddOneWhenAbsent() throws Exception {
+    }
+
+    @Test
+    public void retrieveMailboxCounterShouldWorkWhenFullRow() throws Exception {
```
Maybe it's because of my lack of knowledge about the base code but I do not find the name of this test clear.

chibenwa replied:
Full row => both counter are incremented

Feel free to propose a commit for rename

strokyl commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MailboxFactory.java:
```diff
@@ -93,13 +103,22 @@ public MailboxFactory(MailboxManager mailboxManager) {
     }
 
-    @VisibleForTesting Optional<MailboxId> getParentIdFromMailboxPath(MailboxPath mailboxPath, MailboxSession mailboxSession) throws MailboxException {
+    @VisibleForTesting Optional<MailboxId> getParentIdFromMailboxPath(MailboxPath mailboxPath, Optional<List<MailboxMetaData>> userMailboxesMetadata,
```
For me it seems we could have use `List<MailboxMetaData>` instead of `Optional<List<MailboxMetaData>>` the empty case would be a empty list. No ?

chibenwa replied:
No.

It needs refactoring to do so.

You have a difference between : 
 - I don't took time to retrieve mailboxes
 - There is no mailboxes

strokyl commented on mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxCounters.java:
```diff
@@ -20,12 +20,40 @@
+        }
+
+        public MailboxCounters build() {
+            Preconditions.checkState(count.isPresent(), "count is compulsory");
```
Thank you I just learn a new english word !

mbaechler commented on server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MailboxFactory.java:
```diff
@@ -41,27 +42,51 @@
+            return this;
+        }
+
+        public MailboxBuilder usingListForParentIdCalculation(List<MailboxMetaData> userMailboxesMetadata) {
```
usingPreloadedMailboxesMetadata

strokyl replied:
done

rouazana commented on mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java:
```diff
@@ -152,8 +154,10 @@
      *            the context for this call, not null
      * @throws MailboxException
      *             when creation fails
+     * @return Empty optional when the name is empty. If mailbox is created, the id of the mailboxPath specified as
```
which name?

rouazana commented on mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java:
```diff
@@ -152,8 +154,10 @@
      * @throws MailboxException
      *             when creation fails
+     * @return Empty optional when the name is empty. If mailbox is created, the id of the mailboxPath specified as
+     *  parameter is returned (and not potential mailboxIds of parent mailboxes created in the process will be omitted)
```
s/not//

rouazana commented on server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessorTest.java:
```diff
@@ -20,8 +20,9 @@
 import static org.assertj.core.api.Assertions.assertThat;
-import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.*;
```
no wildcard import please

chibenwa replied:
@strokyl Please review your IDE settings

If intelliJ : ask me to give you my conf, I have no such pb, otherwise ask Lyon people.

chibenwa commented:
Just rebased

I also reworked a bit the indent hell of CassandraMailboxMapper::save indent hell : 
 - Improved indent, and better une lambdas
 - Extract two more methods : trySave and manageException. It should be better for reading...

chibenwa commented:
Merged
